File: vscode/src/core/logs/logParser.ts
import {
  LogDocument,
  LogEntry,
  LogSession,
  NetworkEntry,
  NetworkKind,
  RawLine,
  StandardLine,
} from './logModel';
import { displayTimestamp, parseTimestampMs } from './timestamps';

/**
 * Port of `dev.nylo.plugin.logs.parse.LogParser`.
 *
 * Parses the NDJSON written by Nylo's `NyFileLogger` (one JSON object per line) into a
 * {@link LogDocument}. Pure and IDE-independent. Display text is reconstructed into each entry's
 * `raw`, so downstream filtering/rendering work over readable text, never JSON.
 */

const BANNER_RULE_LEN = 60;

export function parseLog(text: string): LogDocument {
  // Normalize line endings so the model/render stay consistent with pasted Windows content.
  const normalized = text.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
  const lines = normalized.split('\n');
  const n = lines.length > 1 && lines[lines.length - 1] === '' ? lines.length - 1 : lines.length;

  const preamble: LogEntry[] = [];
  const sessions: LogSession[] = [];
  let current: SessionBuilder | null = null;

  for (let i = 0; i < n; i++) {
    const line = lines[i];
    if (line.trim().length === 0) {
      continue; // separator written between sessions
    }

    const obj = parseObject(line);
    const lineNo = i + 1;
    const sink = current ? current.entries : preamble;

    if (obj == null) {
      sink.push(rawLine(line, lineNo));
      continue;
    }

    switch (str(obj, 't')) {
      case 'session':
        if (current) {
          sessions.push(current.build());
        }
        current = new SessionBuilder(obj, lineNo);
        break;
      case 'net':
        (current ? current.entries : preamble).push(networkEntry(obj, lineNo));
        break;
      case 'console':
        (current ? current.entries : preamble).push(standardLine(obj, lineNo, true));
        break;
      case 'log':
        (current ? current.entries : preamble).push(standardLine(obj, lineNo, false));
        break;
      default:
        (current ? current.entries : preamble).push(rawLine(line, lineNo));
        break;
    }
  }
  if (current) {
    sessions.push(current.build());
  }
  return { preamble, sessions };
}

function rawLine(line: string, lineNo: number): RawLine {
  return { entryType: 'raw', raw: line, lineStart: lineNo, lineEnd: lineNo };
}

function standardLine(obj: JsonObject, lineNo: number, console: boolean): StandardLine {
  const tsRaw = str(obj, 'ts') ?? str(obj, 'timestamp');
  const tsDisplay = tsRaw != null ? displayTimestamp(tsRaw) : '';
  const session = str(obj, 'session');
  const level = console ? null : str(obj, 'level');
  const message = str(obj, 'msg') ?? str(obj, 'message') ?? '';
  const contextJson = obj['context'] != null ? JSON.stringify(obj['context']) : null;
  const stack = str(obj, 'stack');

  let sb = '';
  if (tsDisplay.length > 0) {
    sb += `[${tsDisplay}] `;
  }
  if (!console && session != null) {
    sb += `[session:${session}] `;
  }
  if (level != null) {
    sb += `[${level}] `;
  }
  sb += message;
  if (contextJson != null) {
    sb += ` ${contextJson}`;
  }
  if (stack != null) {
    sb += `\n${stack}`;
  }

  return {
    entryType: 'standard',
    timestamp: parseTimestampMs(tsRaw),
    timestampRaw: tsDisplay,
    sessionTag: session,
    level,
    message,
    context: contextJson,
    stack,
    raw: sb,
    lineStart: lineNo,
    lineEnd: lineNo,
  };
}

function networkEntry(obj: JsonObject, lineNo: number): NetworkEntry {
  const kindRaw = (str(obj, 'kind') ?? str(obj, 'type'))?.toLowerCase();
  const netKind: NetworkKind = kindRaw === 'response' ? 'response' : kindRaw === 'error' ? 'error' : 'request';

  const requestId = str(obj, 'requestId');
  const method = str(obj, 'method');
  const uri = str(obj, 'uri');
  const statusCode = int(obj, 'statusCode');
  const statusMessage = str(obj, 'statusMessage');
  const responseTimeMs = int(obj, 'responseTimeMs');
  const payloadSize = str(obj, 'payloadSize');

  const summary = networkSummary(netKind, method, uri, statusCode, statusMessage, responseTimeMs, payloadSize, requestId);

  let sb = summary;
  const errorType = str(obj, 'errorType');
  if (errorType != null) {
    sb += `\n  errorType: ${errorType}`;
  }
  const message = str(obj, 'message');
  if (message != null) {
    sb += `\n  message: ${message}`;
  }
  sb += jsonBlock('headers', obj['headers']);
  sb += jsonBlock('body', obj['body']);
  sb += jsonBlock('data', obj['data']);

  return {
    entryType: 'network',
    netKind,
    requestId,
    method,
    uri,
    statusCode,
    statusMessage,
    responseTimeMs,
    payloadSize,
    summary,
    raw: sb,
    lineStart: lineNo,
    lineEnd: lineNo,
  };
}

function networkSummary(
  kind: NetworkKind,
  method: string | null,
  uri: string | null,
  statusCode: number | null,
  statusMessage: string | null,
  responseTimeMs: number | null,
  payloadSize: string | null,
  requestId: string | null,
): string {
  let symbol: string;
  if (kind === 'request') {
    symbol = '→';
  } else if (kind === 'error') {
    symbol = '✗';
  } else if (statusCode != null && statusCode >= 200 && statusCode <= 299) {
    symbol = '✓';
  } else {
    symbol = '⚠';
  }
  const parts: string[] = [`${symbol} [${method ?? '?'}] ${uri ?? '-'}`];
  const status = [statusCode != null ? String(statusCode) : null, statusMessage].filter((x) => x != null).join(' ').trim();
  if (status.length > 0) {
    parts.push(status);
  }
  if (responseTimeMs != null) {
    parts.push(`${responseTimeMs}ms`);
  }
  if (payloadSize != null) {
    parts.push(payloadSize);
  }
  if (requestId != null) {
    parts.push(`ID ${requestId}`);
  }
  return parts.join('  ·  ');
}

function jsonBlock(label: string, value: unknown): string {
  if (value === undefined || value === null) {
    return '';
  }
  if (isPlainObject(value) && Object.keys(value).length === 0) {
    return '';
  }
  if (Array.isArray(value) && value.length === 0) {
    return '';
  }
  const pretty = JSON.stringify(value, null, 2)
    .split('\n')
    .map((l) => `  ${l}`)
    .join('\n');
  return `\n  ${label}:\n${pretty}`;
}

function bannerLines(id: string, started: string, platform: string, app: string, version: string, env: string): string[] {
  const rule = '='.repeat(BANNER_RULE_LEN);
  return [
    rule,
    ` SESSION  ${id}`,
    ` started   ${started}`,
    ` platform  ${platform}`,
    ` app       ${app}`,
    ` version   ${version}`,
    ` env       ${env}`,
    rule,
  ];
}

class SessionBuilder {
  readonly tag: string;
  readonly fullId: string;
  readonly started: number | null;
  readonly startedRaw: string;
  readonly platform: string;
  readonly app: string;
  readonly version: string;
  readonly env: string;
  readonly bannerLines: string[];
  readonly startLine: number;
  readonly entries: LogEntry[] = [];

  constructor(obj: JsonObject, lineNo: number) {
    this.fullId = str(obj, 'id') ?? '-';
    const startedRaw = str(obj, 'started');
    this.startedRaw = startedRaw != null ? displayTimestamp(startedRaw) : '-';
    this.platform = str(obj, 'platform') ?? '-';
    this.app = str(obj, 'app') ?? '-';
    this.version = str(obj, 'version') ?? '-';
    this.env = str(obj, 'env') ?? '-';
    this.tag = substringAfterLast(this.fullId, '-');
    this.started = parseTimestampMs(startedRaw);
    this.bannerLines = bannerLines(this.fullId, this.startedRaw, this.platform, this.app, this.version, this.env);
    this.startLine = lineNo;
  }

  build(): LogSession {
    const last = this.entries[this.entries.length - 1];
    const end = last ? last.lineEnd : this.startLine;
    return {
      tag: this.tag,
      fullId: this.fullId,
      started: this.started,
      startedRaw: this.startedRaw,
      platform: this.platform,
      app: this.app,
      version: this.version,
      env: this.env,
      bannerLines: this.bannerLines,
      entries: this.entries.slice(),
      lineStart: this.startLine,
      lineEnd: end,
    };
  }
}

// --- JSON helpers tolerant of missing / null / wrong-typed values -------------------------------

type JsonObject = Record<string, unknown>;

function parseObject(line: string): JsonObject | null {
  try {
    const parsed: unknown = JSON.parse(line);
    return isPlainObject(parsed) ? (parsed as JsonObject) : null;
  } catch {
    return null;
  }
}

function isPlainObject(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function str(obj: JsonObject, key: string): string | null {
  const el = obj[key];
  if (el === undefined || el === null) {
    return null;
  }
  let s: string;
  if (typeof el === 'string') {
    s = el;
  } else if (typeof el === 'number' || typeof el === 'boolean') {
    s = String(el);
  } else {
    s = JSON.stringify(el);
  }
  // A JSON-escaped \r survives the whole-file normalization above (in the file it's only the
  // two-char escape). This accessor is the one choke point every decoded string passes through,
  // so normalize here — `raw` must never carry CRs.
  return s.includes('\r') ? s.replace(/\r\n/g, '\n').replace(/\r/g, '\n') : s;
}

function int(obj: JsonObject, key: string): number | null {
  const el = obj[key];
  if (typeof el !== 'number' || !Number.isFinite(el)) {
    return null;
  }
  return Math.trunc(el);
}

function substringAfterLast(value: string, delimiter: string): string {
  const idx = value.lastIndexOf(delimiter);
  return idx < 0 ? value : value.slice(idx + delimiter.length);
}

File: intellij/src/main/kotlin/dev/nylo/plugin/logs/parse/LogParser.kt
package dev.nylo.plugin.logs.parse

import com.google.gson.GsonBuilder
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import dev.nylo.plugin.logs.model.LogDocument
import dev.nylo.plugin.logs.model.LogEntry
import dev.nylo.plugin.logs.model.LogSession
import dev.nylo.plugin.logs.model.NetworkEntry
import dev.nylo.plugin.logs.model.RawLine
import dev.nylo.plugin.logs.model.StandardLine
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter

/**
 * Parses the NDJSON on-disk format written by Nylo's `NyFileLogger` (one JSON object per line) into
 * a [LogDocument]. Pure and IDE-independent so it can be unit-tested directly.
 *
 * Record shapes (authoritative source: `support/lib/helpers/src/ny_file_logger.dart`):
 *  - `{"t":"session", id, started, platform, app, version, env}` — starts a [LogSession]; the parser
 *    reconstructs the familiar `====` banner into [LogSession.bannerLines] for display.
 *  - `{"t":"log", ts, session, level?, msg, context?, stack?}` — a [StandardLine].
 *  - `{"t":"console", ts, session, msg}` — a [StandardLine] with a null level.
 *  - `{"t":"net", kind, ts, session, requestId, method, uri, statusCode?, ...}` — a [NetworkEntry].
 *
 * Each record occupies exactly one physical line. Display text is reconstructed here into each
 * entry's [LogEntry.raw], so downstream filtering/rendering work over readable text, never JSON.
 */
object LogParser {

    private val GSON = GsonBuilder().disableHtmlEscaping().create()
    private val PRETTY = GsonBuilder().setPrettyPrinting().disableHtmlEscaping().create()
    private val TS_DISPLAY = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
    private val TRAILING_UTC_OFFSET = Regex("""[+-]\d{2}(:?\d{2})?$""")
    private const val BANNER_RULE_LEN = 60

    fun parse(text: CharSequence): LogDocument {
        // Normalize line endings: IntelliJ Documents reject \r\n / \r, and pasted bodies sometimes carry
        // Windows endings. Doing it here keeps the model and render consistent.
        val normalized = text.toString().replace("\r\n", "\n").replace('\r', '\n')
        val lines = normalized.split('\n')
        val n = if (lines.size > 1 && lines.last().isEmpty()) lines.size - 1 else lines.size

        val preamble = mutableListOf<LogEntry>()
        val sessions = mutableListOf<LogSession>()
        var current: SessionBuilder? = null

        for (i in 0 until n) {
            val line = lines[i]
            if (line.isBlank()) continue // separator written between sessions

            val obj = runCatching { JsonParser.parseString(line) }.getOrNull()
                ?.takeIf { it.isJsonObject }?.asJsonObject
            val lineNo = i + 1

            if (obj == null) {
                (current?.entries ?: preamble).add(RawLine(line, lineNo..lineNo))
                continue
            }

            when (obj.str("t")) {
                "session" -> {
                    current?.let { sessions.add(it.build()) }
                    current = sessionBuilder(obj, lineNo)
                }
                "net" -> (current?.entries ?: preamble).add(networkEntry(obj, lineNo))
                "console" -> (current?.entries ?: preamble).add(standardLine(obj, lineNo, console = true))
                "log" -> (current?.entries ?: preamble).add(standardLine(obj, lineNo, console = false))
                else -> (current?.entries ?: preamble).add(RawLine(line, lineNo..lineNo))
            }
        }
        current?.let { sessions.add(it.build()) }
        return LogDocument(preamble, sessions)
    }

    private fun sessionBuilder(obj: JsonObject, lineNo: Int): SessionBuilder {
        val fullId = obj.str("id") ?: "-"
        val startedRaw = obj.str("started")?.let { displayTimestamp(it) } ?: "-"
        val platform = obj.str("platform") ?: "-"
        val app = obj.str("app") ?: "-"
        val version = obj.str("version") ?: "-"
        val env = obj.str("env") ?: "-"
        return SessionBuilder(
            tag = fullId.substringAfterLast('-'),
            fullId = fullId,
            started = parseTimestamp(obj.str("started")),
            startedRaw = startedRaw,
            platform = platform,
            app = app,
            version = version,
            env = env,
            bannerLines = bannerLines(fullId, startedRaw, platform, app, version, env),
            startLine = lineNo,
        )
    }

    private fun standardLine(obj: JsonObject, lineNo: Int, console: Boolean): StandardLine {
        val tsRaw = obj.str("ts") ?: obj.str("timestamp")
        val tsDisplay = tsRaw?.let { displayTimestamp(it) } ?: ""
        val session = obj.str("session")
        val level = if (console) null else obj.str("level")
        val message = (obj.str("msg") ?: obj.str("message")).orEmpty()
        val contextJson = obj.obj("context")?.let { GSON.toJson(it) }
        val stack = obj.str("stack")

        val sb = StringBuilder()
        if (tsDisplay.isNotEmpty()) sb.append('[').append(tsDisplay).append("] ")
        if (!console && session != null) sb.append("[session:").append(session).append("] ")
        if (level != null) sb.append('[').append(level).append("] ")
        sb.append(message)
        if (contextJson != null) sb.append(' ').append(contextJson)
        if (stack != null) sb.append('\n').append(stack)

        return StandardLine(
            timestamp = parseTimestamp(tsRaw),
            timestampRaw = tsDisplay,
            sessionTag = session,
            level = level,
            message = message,
            context = contextJson,
            stack = stack,
            raw = sb.toString(),
            lineRange = lineNo..lineNo,
        )
    }

    private fun networkEntry(obj: JsonObject, lineNo: Int): NetworkEntry {
        val kind = when ((obj.str("kind") ?: obj.str("type"))?.lowercase()) {
            "request" -> NetworkEntry.Kind.REQUEST
            "response" -> NetworkEntry.Kind.RESPONSE
            "error" -> NetworkEntry.Kind.ERROR
            else -> NetworkEntry.Kind.REQUEST
        }
        val requestId = obj.str("requestId")
        val method = obj.str("method")
        val uri = obj.str("uri")
        val statusCode = obj.int("statusCode")
        val statusMessage = obj.str("statusMessage")
        val responseTimeMs = obj.long("responseTimeMs")
        val payloadSize = obj.str("payloadSize")

        val summary = networkSummary(kind, method, uri, statusCode, statusMessage, responseTimeMs, payloadSize, requestId)

        // Expanded detail: the summary followed by any structured headers/body the record carried.
        val sb = StringBuilder(summary)
        obj.str("errorType")?.let { sb.append("\n  errorType: ").append(it) }
        obj.str("message")?.let { sb.append("\n  message: ").append(it) }
        appendJsonBlock(sb, "headers", obj.obj("headers"))
        appendJsonBlock(sb, "body", obj.obj("body"))
        appendJsonBlock(sb, "data", obj.obj("data"))

        return NetworkEntry(
            kind = kind,
            requestId = requestId,
            method = method,
            uri = uri,
            statusCode = statusCode,
            statusMessage = statusMessage,
            responseTimeMs = responseTimeMs,
            payloadSize = payloadSize,
            summary = summary,
            raw = sb.toString(),
            lineRange = lineNo..lineNo,
        )
    }

    private fun networkSummary(
        kind: NetworkEntry.Kind,
        method: String?,
        uri: String?,
        statusCode: Int?,
        statusMessage: String?,
        responseTimeMs: Long?,
        payloadSize: String?,
        requestId: String?,
    ): String {
        val symbol = when {
            kind == NetworkEntry.Kind.REQUEST -> "→"
            kind == NetworkEntry.Kind.ERROR -> "✗"
            statusCode != null && statusCode in 200..299 -> "✓"
            else -> "⚠"
        }
        val parts = mutableListOf<String>()
        parts += "$symbol [${method ?: "?"}] ${uri ?: "-"}"
        val status = listOfNotNull(statusCode?.toString(), statusMessage).joinToString(" ").trim()
        if (status.isNotEmpty()) parts += status
        responseTimeMs?.let { parts += "${it}ms" }
        payloadSize?.let { parts += it }
        requestId?.let { parts += "ID $it" }
        return parts.joinToString("  ·  ")
    }

    private fun appendJsonBlock(sb: StringBuilder, label: String, element: JsonElement?) {
        if (element == null) return
        if (element.isJsonObject && element.asJsonObject.size() == 0) return
        if (element.isJsonArray && element.asJsonArray.size() == 0) return
        val pretty = PRETTY.toJson(element).prependIndent("  ")
        sb.append("\n  ").append(label).append(':').append('\n').append(pretty)
    }

    private fun bannerLines(
        id: String,
        started: String,
        platform: String,
        app: String,
        version: String,
        env: String,
    ): List<String> {
        val rule = "=".repeat(BANNER_RULE_LEN)
        return listOf(
            rule,
            " SESSION  $id",
            " started   $started",
            " platform  $platform",
            " app       $app",
            " version   $version",
            " env       $env",
            rule,
        )
    }

    private fun displayTimestamp(raw: String): String =
        parseTimestamp(raw)?.format(TS_DISPLAY) ?: raw

    private fun parseTimestamp(raw: String?): LocalDateTime? {
        if (raw == null) return null
        // Accept ISO-8601 ("2026-06-26T10:32:59[.sss][Z|±HH:mm]") and the legacy
        // "yyyy-MM-dd HH:mm:ss" form. Offsets are dropped, not converted — the wall-clock time as
        // logged is what should display and sort. A plain substringBefore('+') would miss negative
        // offsets (the '-' also appears in the date), so strip a trailing zone suffix by pattern.
        val core = raw.trim().replace(' ', 'T')
        return runCatching { LocalDateTime.parse(core) }.getOrNull()
            ?: runCatching { LocalDateTime.parse(core.removeSuffix("Z").replace(TRAILING_UTC_OFFSET, "")) }.getOrNull()
    }

    // --- gson field accessors that tolerate missing / null / wrong-typed values -------------------

    private fun JsonObject.str(key: String): String? {
        val el = get(key) ?: return null
        if (el.isJsonNull) return null
        val s = if (el.isJsonPrimitive) el.asString else el.toString()
        // A JSON-escaped \r survives the whole-file normalization above (it exists in the file only
        // as the two-char escape) and IntelliJ Documents assert on literal CRs. This accessor is the
        // one choke point every decoded string passes through, so normalize here.
        return if ('\r' in s) s.replace("\r\n", "\n").replace('\r', '\n') else s
    }

    private fun JsonObject.int(key: String): Int? = number(key)?.toInt()

    private fun JsonObject.long(key: String): Long? = number(key)?.toLong()

    private fun JsonObject.number(key: String): Double? {
        val el = get(key) ?: return null
        if (!el.isJsonPrimitive || !el.asJsonPrimitive.isNumber) return null
        return runCatching { el.asDouble }.getOrNull()
    }

    private fun JsonObject.obj(key: String): JsonElement? = get(key)?.takeIf { !it.isJsonNull }

    private class SessionBuilder(
        val tag: String,
        val fullId: String,
        val started: LocalDateTime?,
        val startedRaw: String,
        val platform: String,
        val app: String,
        val version: String,
        val env: String,
        val bannerLines: List<String>,
        val startLine: Int,
    ) {
        val entries = mutableListOf<LogEntry>()

        fun build(): LogSession {
            val end = entries.lastOrNull()?.lineRange?.last ?: startLine
            return LogSession(
                tag, fullId, started, startedRaw, platform, app, version, env,
                bannerLines, entries.toList(), startLine..end,
            )
        }
    }
}
